File: carousel.nim
import ../kryon/src/kryon
import std/[math, tables, json, os, random, strutils]

# Constants
const
  MAX_OPTIONS = 20
  CAROUSEL_FILE = "~/Documents/quest/carousel.json"

# Data structures
type
  Theme = object
    name: string
    background: string
    primary: string
    secondary: string
    text: string
    accent: string

  CarouselOption = object
    text: string
    r, g, b: uint8

  CarouselData = object
    options: seq[CarouselOption]
    currentAngle: float

# Global state
var
  options: seq[CarouselOption] = @[]
  currentAngle: float = -90.0
  isSpinning: bool = false
  spinSpeed: float = 0.0
  inputText: string = ""
  isEditMode: bool = false
  editIndex: int = -1
  currentThemeIndex: int = 0

# Theme definitions
let themes = @[
  Theme(
    name: "Dark",
    background: "#1a1a1a",
    primary: "#2d2d2d",
    secondary: "#3d3d3d",
    text: "#ffffff",
    accent: "#4a90e2"
  ),
  Theme(
    name: "Light",
    background: "#f5f5f5",
    primary: "#ffffff",
    secondary: "#e0e0e0",
    text: "#333333",
    accent: "#2196f3"
  ),
  Theme(
    name: "Ocean",
    background: "#0d1b2a",
    primary: "#1b263b",
    secondary: "#415a77",
    text: "#e0e1dd",
    accent: "#778da9"
  )
]

proc getCurrentTheme(): Theme =
  themes[currentThemeIndex]

# Generate random color for options
proc generateRandomColor(): tuple[r, g, b: uint8] =
  result.r = uint8(rand(50..255))
  result.g = uint8(rand(50..255))
  result.b = uint8(rand(50..255))

# Save options to JSON file
proc saveOptions() =
  let filePath = expandTilde(CAROUSEL_FILE)
  let dir = parentDir(filePath)

  if not dirExists(dir):
    createDir(dir)

  var optionsArray = newJArray()
  for opt in options:
    optionsArray.add(%* {
      "text": opt.text,
      "r": opt.r,
      "g": opt.g,
      "b": opt.b
    })

  let jsonNode = %* {
    "options": optionsArray,
    "currentAngle": currentAngle
  }

  try:
    writeFile(filePath, jsonNode.pretty())
  except:
    echo "Error saving carousel: ", getCurrentExceptionMsg()

# Load options from JSON file
proc loadOptions() =
  let filePath = expandTilde(CAROUSEL_FILE)

  if not fileExists(filePath):
    # Create default options
    let color1 = generateRandomColor()
    let color2 = generateRandomColor()
    let color3 = generateRandomColor()
    options = @[
      CarouselOption(text: "Option 1", r: color1.r, g: color1.g, b: color1.b),
      CarouselOption(text: "Option 2", r: color2.r, g: color2.g, b: color2.b),
      CarouselOption(text: "Option 3", r: color3.r, g: color3.g, b: color3.b)
    ]
    return

  try:
    let jsonContent = readFile(filePath)
    let jsonNode = parseJson(jsonContent)

    options = @[]
    for optNode in jsonNode["options"]:
      options.add(CarouselOption(
        text: optNode["text"].getStr(),
        r: uint8(optNode["r"].getInt()),
        g: uint8(optNode["g"].getInt()),
        b: uint8(optNode["b"].getInt())
      ))

    if jsonNode.hasKey("currentAngle"):
      currentAngle = jsonNode["currentAngle"].getFloat()
  except:
    echo "Error loading carousel: ", getCurrentExceptionMsg()
    let color1 = generateRandomColor()
    options = @[CarouselOption(text: "Option 1", r: color1.r, g: color1.g, b: color1.b)]

# Get sector angle
proc getSectorAngle(): float =
  if options.len == 0:
    return 0.0
  360.0 / options.len.float

# Update spinning animation
proc updateSpin() =
  if not isSpinning:
    return

  if spinSpeed > 1.0:
    currentAngle += spinSpeed
    spinSpeed *= 0.97
  else:
    # Snap to nearest sector
    let sectorAngle = getSectorAngle()
    let halfSectorAngle = sectorAngle / 2.0

    let normalizedAngle = currentAngle + halfSectorAngle
    let sector = floor(normalizedAngle / sectorAngle)
    let targetAngle = (sector * sectorAngle) - halfSectorAngle

    var angleDiff = targetAngle - currentAngle
    if angleDiff > 180.0: angleDiff -= 360.0
    if angleDiff < -180.0: angleDiff += 360.0

    let moveSpeed = angleDiff * 0.1
    currentAngle += moveSpeed

    if abs(moveSpeed) < 0.01:
      isSpinning = false
      currentAngle = targetAngle
      spinSpeed = 0.0
      saveOptions()

  # Normalize angle
  while currentAngle >= 360.0: currentAngle -= 360.0
  while currentAngle < 0.0: currentAngle += 360.0

# Draw the carousel wheel
proc drawCarousel(ctx: DrawingContext, width, height: float) =
  if options.len == 0:
    return

  let theme = getCurrentTheme()
  let centerX = width / 2.0
  let centerY = height / 2.0
  let radius = min(width, height) * 0.4
  let sectorAngle = getSectorAngle()

  # Draw sectors
  for i in 0..<options.len:
    let startAngle = currentAngle + i.float * sectorAngle - 90.0
    let endAngle = startAngle + sectorAngle

    # Convert to radians for arc drawing
    let startRad = startAngle * PI / 180.0
    let endRad = endAngle * PI / 180.0

    # Blend option color with theme
    let opt = options[i]
    let blendedR = (opt.r.int + parseHexInt(theme.accent[1..2])) div 2
    let blendedG = (opt.g.int + parseHexInt(theme.accent[3..4])) div 2
    let blendedB = (opt.b.int + parseHexInt(theme.accent[5..6])) div 2

    ctx.fillStyle = rgba(blendedR.uint8, blendedG.uint8, blendedB.uint8, 255)

    # Draw sector using arc
    ctx.beginPath()
    ctx.moveTo(centerX, centerY)
    ctx.arc(centerX, centerY, radius, startRad, endRad)
    ctx.closePath()
    ctx.fill()

    # Draw sector outline
    ctx.strokeStyle = rgba(0, 0, 0, 100)
    ctx.lineWidth = 2.0
    ctx.beginPath()
    ctx.moveTo(centerX, centerY)
    ctx.arc(centerX, centerY, radius, startRad, endRad)
    ctx.closePath()
    ctx.stroke()

  # Draw pointer triangle
  let accentColor = theme.accent
  let accentR = parseHexInt(accentColor[1..2]).uint8
  let accentG = parseHexInt(accentColor[3..4]).uint8
  let accentB = parseHexInt(accentColor[5..6]).uint8

  ctx.fillStyle = rgba(accentR, accentG, accentB, 255)
  ctx.beginPath()
  ctx.moveTo(centerX, centerY - radius - 20)
  ctx.lineTo(centerX - 10, centerY - radius)
  ctx.lineTo(centerX + 10, centerY - radius)
  ctx.closePath()
  ctx.fill()

# Add new option
proc addOption() =
  if options.len >= MAX_OPTIONS:
    return

  if inputText.len == 0:
    return

  if isEditMode and editIndex >= 0 and editIndex < options.len:
    # Edit existing
    options[editIndex].text = inputText
    isEditMode = false
    editIndex = -1
  else:
    # Add new
    let color = generateRandomColor()
    options.add(CarouselOption(text: inputText, r: color.r, g: color.g, b: color.b))

  inputText = ""
  saveOptions()

# Delete option
proc deleteOption(index: int) =
  if index >= 0 and index < options.len:
    options.delete(index)
    saveOptions()

# Edit option
proc editOption(index: int) =
  if index >= 0 and index < options.len:
    inputText = options[index].text
    isEditMode = true
    editIndex = index
    echo isEditMode

# Spin the wheel
proc spinWheel() =
  if options.len == 0:
    return

  isSpinning = true
  spinSpeed = 20.0 + rand(10.0)
  # Don't save here - save when spin completes

# Clear all options
proc clearAllOptions() =
  options = @[]
  isSpinning = false
  spinSpeed = 0.0
  currentAngle = -90.0
  saveOptions()

# Initialize
randomize()
loadOptions()

# Main application
let app = kryonApp:
  Header:
    width = 900
    height = 700
    title = "Carousel"

  Body:
    backgroundColor = getCurrentTheme().background


    Text:
      text = "Decision Wheel"
      fontSize = 32
      color = getCurrentTheme().text

    Row:
      gap = 10
      alignItems = "center"
      Text:
        text = "Theme:"
        fontSize = 16
        color = getCurrentTheme().text

      for i in 0..<themes.len:
        Button:
          text = themes[i].name
          onClick = proc() = currentThemeIndex = i
          backgroundColor = if i == currentThemeIndex: getCurrentTheme().accent else: getCurrentTheme().secondary
          textColor = getCurrentTheme().text
          width = 80
          height = 30
          fontSize = 14

    # Left side - Options list
    Column:
      width = 350
      gap = 10
      backgroundColor = getCurrentTheme().primary

      for i in 0..<options.len:
        Row:
          gap = 10
          alignItems = "center"

          Text:
            text = $(i + 1) & ". " & options[i].text
            fontSize = 16
            color = getCurrentTheme().text
            width = 180

          Button:
            text = "Edit"
            onClick = editOption(i)
            backgroundColor = getCurrentTheme().accent
            textColor = getCurrentTheme().text
            width = 60
            height = 25
            fontSize = 12

          Button:
            text = "Delete"
            onClick = deleteOption(i)
            backgroundColor = "#d32f2f"
            textColor = "#ffffff"
            width = 60
            height = 25
            fontSize = 12

      Spacer()
      
      Text:
        echo isEditMode
        text = if isEditMode: "Editing option..." else: "Add new option"
        fontSize = 14
        color = getCurrentTheme().text

      Input:
        value = inputText
        onTextChange = proc(text: string) = inputText = text
        onSubmit = proc() = addOption()
        fontSize = 16
        color = getCurrentTheme().text
        backgroundColor = getCurrentTheme().secondary
        width = 350
        height = 40

      Button:
        text = if isEditMode: "Save Edit" else: "Add Option"
        onClick = proc() = addOption()
        backgroundColor = getCurrentTheme().accent
        textColor = getCurrentTheme().text
        width = 350
        height = 40
        fontSize = 16

File: carousel.lua
-- Decision Wheel Carousel
-- A spinning wheel app built with Kryon and Canvas plugin

local UI = require("kryon.dsl")
local Reactive = require("kryon.reactive")
local canvas = require("canvas")

-- ============================================================================
-- App State
-- ============================================================================

local state = {
    options = {
        {text = "Meditation", r = 100, g = 180, b = 255},  -- Blue
        {text = "Push ups", r = 255, g = 100, b = 100},    -- Red
        {text = "Jogging", r = 100, g = 255, b = 100},     -- Green
    },
    currentAngle = 0,
    spinSpeed = 0,
    isSpinning = false,
    inputText = "",
    editIndex = -1,
    currentTheme = 1,
}

local themes = {
    {
        name = "Dark",
        background = "#0a0a0a",
        primary = "#151515",
        secondary = "#202020",
        text = "#ffffff",
        accent = "#4a9eff",
    },
    {
        name = "Light",
        background = "#f0f0f0",
        primary = "#ffffff",
        secondary = "#e0e0e0",
        text = "#000000",
        accent = "#4a9eff",
    },
    {
        name = "Ocean",
        background = "#0a192f",
        primary = "#112240",
        secondary = "#1d3557",
        text = "#8892b0",
        accent = "#64ffda",
    },
}

-- ============================================================================
-- Utility Functions
-- ============================================================================

local function colorToRGBA(r, g, b, a)
    a = a or 255
    return bit.lshift(r, 24) + bit.lshift(g, 16) + bit.lshift(b, 8) + a
end

local function hexToRGBA(hex)
    hex = hex:gsub("#", "")
    local r = tonumber(hex:sub(1, 2), 16)
    local g = tonumber(hex:sub(3, 4), 16)
    local b = tonumber(hex:sub(5, 6), 16)
    local a = 255
    if #hex == 8 then
        a = tonumber(hex:sub(7, 8), 16)
    end
    return colorToRGBA(r, g, b, a)
end

local function blendColors(r1, g1, b1, r2, g2, b2, blend)
    local r = math.floor(r1 * (1 - blend) + r2 * blend)
    local g = math.floor(g1 * (1 - blend) + g2 * blend)
    local b = math.floor(b1 * (1 - blend) + b2 * blend)
    return r, g, b
end

-- ============================================================================
-- Carousel Drawing
-- ============================================================================

local function drawCarousel(canvasWidth, canvasHeight)
    print(string.format("[CAROUSEL] drawCarousel called: width=%d height=%d, options=%d", canvasWidth, canvasHeight, #state.options))

    -- CRITICAL: Reset transform matrix to identity at start of each frame
    -- Without this, transforms accumulate across frames causing rightward drift
    canvas.origin()

    local theme = themes[state.currentTheme]

    -- Clear canvas with background color
    local bgHex = theme.secondary:gsub("#", "")
    local bgR = tonumber(bgHex:sub(1, 2), 16)
    local bgG = tonumber(bgHex:sub(3, 4), 16)
    local bgB = tonumber(bgHex:sub(5, 6), 16)
    canvas.clear(bgR, bgG, bgB, 255)

    local centerX = canvasWidth / 2
    local centerY = canvasHeight / 2
    -- Use 80px padding to ensure wheel fits comfortably with even spacing
    local radius = math.min(canvasWidth, canvasHeight) / 2 - 80
    print(string.format("[CAROUSEL] Center: (%.1f, %.1f), Radius: %.1f", centerX, centerY, radius))

    if #state.options == 0 then
        print("[CAROUSEL] No options to draw")
        return
    end
    local angleStep = (2 * math.pi) / #state.options
    local currentAngleRad = math.rad(state.currentAngle)

    -- Draw wheel sectors as pie chart
    for i, option in ipairs(state.options) do
        local startAngle = (i - 1) * angleStep + currentAngleRad - math.pi / 2
        local endAngle = i * angleStep + currentAngleRad - math.pi / 2

        -- Blend option color with theme accent for visual appeal
        local accentR, accentG, accentB = 0x4a, 0x9e, 0xff  -- Default blue accent
        if theme.accent then
            local accentHex = theme.accent:gsub("#", "")
            accentR = tonumber(accentHex:sub(1, 2), 16) or accentR
            accentG = tonumber(accentHex:sub(3, 4), 16) or accentG
            accentB = tonumber(accentHex:sub(5, 6), 16) or accentB
        end

        local r, g, b = blendColors(option.r, option.g, option.b, accentR, accentG, accentB, 0.2)

        -- Draw pie sector as polygon (arc doesn't support fill mode)
        canvas.fill(r, g, b)

        -- Create polygon vertices for pie sector
        local vertices = {centerX, centerY}  -- Start at center
        local numSegments = 32
        for j = 0, numSegments do
            local angle = startAngle + (endAngle - startAngle) * (j / numSegments)
            local px = centerX + math.cos(angle) * radius
            local py = centerY + math.sin(angle) * radius
            table.insert(vertices, px)
            table.insert(vertices, py)
        end
        canvas.polygon("fill", vertices)

        -- Draw text on sector
        local midAngle = (startAngle + endAngle) / 2
        local textRadius = radius * 0.6
        local textX = centerX + math.cos(midAngle) * textRadius
        local textY = centerY + math.sin(midAngle) * textRadius

        -- Measure text to center it properly
        local textWidth = canvas.getTextWidth(option.text)
        local textHeight = canvas.getTextHeight()

        -- Save transform state
        canvas.push()

        -- Move to text position
        canvas.translate(textX, textY)

        -- Rotate text to align with sector (optional - makes it more readable)
        canvas.rotate(midAngle + math.pi / 2)

        -- Draw centered text
        canvas.fill(255, 255, 255)
        canvas.print(option.text, -textWidth / 2, -textHeight / 2)

        -- Restore transform state
        canvas.pop()
    end

    -- Draw center circle
    if theme.primary then
        local hexColor = theme.primary:gsub("#", "")
        local pr = tonumber(hexColor:sub(1, 2), 16) or 45
        local pg = tonumber(hexColor:sub(3, 4), 16) or 45
        local pb = tonumber(hexColor:sub(5, 6), 16) or 45
        canvas.fill(pr, pg, pb)
    else
        canvas.fill(45, 45, 45)
    end
    canvas.circle("fill", centerX, centerY, 50)

    -- Draw pointer triangle at top
    canvas.fill(255, 0, 0)
    local pointerY = centerY - radius - 10
    canvas.polygon("fill", {
        centerX, pointerY - 10,
        centerX - 10, pointerY + 10,
        centerX + 10, pointerY + 10
    })
end

-- ============================================================================
-- Spin Logic
-- ============================================================================

local function spinWheel()
    if state.isSpinning then
        return
    end

    state.isSpinning = true
    state.spinSpeed = math.random(20, 40)
end

local function updateSpin(deltaTime)
    if not state.isSpinning then
        return
    end

    state.currentAngle = state.currentAngle + state.spinSpeed
    state.spinSpeed = state.spinSpeed * 0.97

    if state.spinSpeed < 0.1 then
        state.isSpinning = false
        state.spinSpeed = 0

        -- Snap to nearest option
        local angleStep = 360 / #state.options
        local normalizedAngle = state.currentAngle % 360
        local nearestIndex = math.floor((normalizedAngle + angleStep / 2) / angleStep) + 1
        state.currentAngle = (nearestIndex - 1) * angleStep
    end
end

-- ============================================================================
-- Options Management
-- ============================================================================

local function addOption()
    if state.inputText ~= "" then
        local r = math.random(100, 255)
        local g = math.random(100, 255)
        local b = math.random(100, 255)

        if state.editIndex >= 0 then
            state.options[state.editIndex].text = state.inputText
            state.editIndex = -1
        else
            table.insert(state.options, {text = state.inputText, r = r, g = g, b = b})
        end

        state.inputText = ""
    end
end

local function deleteOption(index)
    if #state.options > 1 then
        table.remove(state.options, index)
    end
end

local function editOption(index)
    state.inputText = state.options[index].text
    state.editIndex = index
end

-- ============================================================================
-- UI Construction
-- ============================================================================

local theme = themes[state.currentTheme]

local root = UI.Column {
    width = "1200px",
    height = "900px",
    background = theme.background,
    windowTitle = "Decision Wheel",
    children = {
        -- Header
        UI.Container {
            width = "100%",
            height = "60px",
            background = theme.primary,
            padding = "10px",
            children = {
                UI.Text {
                    text = "Decision Wheel",
                    fontSize = 32,
                    color = theme.text,
                },
            },
        },

        -- Main content
        UI.Row {
            width = "100%",
            height = "840px",
            gap = 20,
            padding = "20px",
            children = {
                -- Options panel
                UI.Column {
                    width = "300px",
                    height = "100%",
                    gap = 10,
                    background = theme.primary,
                    padding = "15px",
                    children = {
                        UI.Text {
                            text = "Options",
                            fontSize = 20,
                            color = theme.text,
                        },

                        -- Input field
                        UI.Input {
                            placeholder = "Enter option...",
                            value = state.inputText,
                            onChange = function(value)
                                state.inputText = value
                            end,
                        },

                        -- Add button
                        UI.Button {
                            text = state.editIndex >= 0 and "Update" or "Add Option",
                            onClick = addOption,
                            background = theme.accent,
                            color = "#ffffff",
                            padding = 10,
                            width = "100%",
                        },

                        -- Options list
                        UI.Column {
                            gap = 5,
                            children = {
                                -- Option 1: Meditation
                                UI.Row {
                                    width = "100%",
                                    gap = 5,
                                    padding = "8px",
                                    background = "#2a2a2a",
                                    children = {
                                        UI.Text {
                                            text = "Meditation",
                                            color = theme.text,
                                            fontSize = 14,
                                        },
                                    },
                                },

                                -- Option 2: Push ups
                                UI.Row {
                                    width = "100%",
                                    gap = 5,
                                    padding = "8px",
                                    background = "#2a2a2a",
                                    children = {
                                        UI.Text {
                                            text = "Push ups",
                                            color = theme.text,
                                            fontSize = 14,
                                        },
                                    },
                                },

                                -- Option 3: Jogging
                                UI.Row {
                                    width = "100%",
                                    gap = 5,
                                    padding = "8px",
                                    background = "#2a2a2a",
                                    children = {
                                        UI.Text {
                                            text = "Jogging",
                                            color = theme.text,
                                            fontSize = 14,
                                        },
                                    },
                                },
                            },
                        },
                    },
                },

                -- Canvas panel
                UI.Column {
                    width = "850px",
                    height = "100%",
                    gap = 10,
                    alignItems = "center",
                    children = {
                        -- Canvas
                        UI.Canvas {
                            width = 800,
                            height = 700,
                            background = theme.secondary,
                            onDraw = function()
                                print("[carousel] Canvas onDraw called")
                                drawCarousel(800, 700)
                            end,
                            onUpdate = function(dt)
                                updateSpin(dt)
                            end,
                        },

                        -- Spin button
                        UI.Button {
                            text = "SPIN!",
                            width = "200px",
                            height = "50px",
                            fontSize = 24,
                            background = theme.accent,
                            color = "#ffffff",
                            onClick = spinWheel,
                        },
                    },
                },
            },
        },
    },
}

-- ============================================================================
-- Run App
-- ============================================================================

-- Create app configuration
local app = {
    root = root,
    window = {
        width = 1200,
        height = 900,
        title = "Decision Wheel"
    }
}

return app
